gh-85747: Active voice & suggested edits, 'running/stopping loop' & 'callbacks' subsections of asyncio-eventloop.rst (#100270)
python/cpython: Doc/library/asyncio-eventloop.rst

.. coroutinemethod:: loop.shutdown_default_executor(timeout=None)

   Schedule the closure of the default executor and wait for it to join all of
   the threads in the :class:`~concurrent.futures.ThreadPoolExecutor`.
   After calling this method,
   :meth:`loop.run_in_executor` will raise a :exc:`RuntimeError` if called

>    while using the default executor.

Reviewer: ```suggestion
   Once this method has been called,
   using the default executor with :meth:`loop.run_in_executor`
   will raise a :exc:`RuntimeError`.
```

This reads a fair bit more clearly to me, as the grammar and flow of ideas is more straightforward for me to parse.

Reviewer 2: I prefer CAM version.

>    The *timeout* parameter specifies the duration in seconds (as a :class:`float`)
>    this method gives the executor to finish joining.

Reviewer: ```suggestion
   The *timeout* parameter specifies the amount of time
   (in :class:`float` seconds) the executor will be given to finish joining.
```

I had trouble getting active voice to not sound awkward here, and I don't think it really adds much to reader understanding here—as a parameter description in an the API reference, the method being described is the clear actor in this case. Therefore, I reverted to the original wording, with the key clarification that the time given is in (floating point) seconds.

Reviewer 2: Float seconds is critical.  I might replace 'will be given' to 'will have', which removes the passive.

Reviewer: We might, but the current framing is consistent with this being an external, discretionary timeout acting on the executor, as opposed to a requirement intrinsic to the executor or its context.

>    With the default, ``None``, the
>    method allows the executor an unlimited amount of time.

Reviewer: ```suggestion
   With the default, ``None``,
   the executor is allowed an unlimited amount of time.
```

Likewise, here, I feel inserting "the method" here makes the sentence structure more complex for the reader to parse without really aiding meaning. Here, the exact implementation detail doing the allowing (the subject) is a lot less important than the object being allowed (the executor), which is exactly what passive voice is useful for. And the method is explicitly what we are describing here, which the basic structure of API reference documentation already makes clear, and repeatedly restating is not really in line with the concision and directness that serves reference-type docs well.

Reviewer 2: or 'the executor has ....

Reviewer: Like the above, that framing seems more appropriate for an intrinsic requirement than an externally-set  discretionary deadline. But if we do change the above, we should presumably change this to match

>    If the *timeout* is reached, this method emits a :exc:`RuntimeWarning` and
>    terminates the default executor without waiting for its threads to finish
>    joining.

Reviewer: ```suggestion
   If the *timeout* is reached, a :exc:`RuntimeWarning` is emitted
   and the default executor is terminated
   without waiting for its threads to finish joining.
```

Like before, reading and thinking over this again, passive voice is generally idiomatic and keeping with the Diataxis principles in API reference documentation in the case of referring to the construct that is explicitly stated as being described, and avoids implying implementation details that may or may not be entirely correct (e.g. it implies the method directly emits the `RuntimeWarning`, when it may in fact be emitted deeper in the execution stack).

Reviewer 2: I was wondering about the implied implementation detail.  agree with leave it out.

   .. note::

>       There is no need to call this method when :func:`asyncio.run` is
>       used, because that high-level function handles default executor shutdown
>       automatically.

Reviewer: ```suggestion
      There is no need to call this method when :func:`asyncio.run` is used,
      because that function handles default executor shutdown automatically.
```

You might want to consider eliding "high-level" to improve readability here and avoid sounding a bit awkward.

Reviewer 2: See my alternate suggestion in my review.

Reviewer: Yeah, I think I like a slight variant of yours better.

Just FYI, you can also reply with a suggestion of your own in the same thread, to keep things in one place.

Reviewer: Applied the minor refinement of your suggestion instead

>       There is no need to call this method when :func:`asyncio.run` is
>       used, because that high-level function handles default executor shutdown
>       automatically.

Reviewer 2: ```suggestion
      Do not call this method when using :func:`asyncio.run`.  The latter handles default executor shutdown
      automatically.
```

Reviewer: I was trying to avoid "the latter" to avoid any confusion, but I think it works okay and flows better overall merged into a single sentence:

```suggestion
      Do not call this method when using :func:`asyncio.run`,
      as the latter handles default executor shutdown automatically.
```

.. coroutinemethod:: loop.shutdown_default_executor(timeout=None)

   Schedule the closure of the default executor and wait for it to join all of

>    the threads in the :class:`ThreadPoolExecutor`. After calling this method,
>    :meth:`loop.run_in_executor` will raise a :exc:`RuntimeError` if called

Reviewer: ```suggestion
   the threads in the :class:`~concurrent.futures.ThreadPoolExecutor`.
   After calling this method,
   :meth:`loop.run_in_executor` will raise a :exc:`RuntimeError` if called
```

Fix broken reference to `ThreadPoolExecutor` (while still displaying it as such)

   while using the default executor.

>    The *timeout* parameter specifies the amount of time this method gives the
>    executor to finish joining. The default value is ``None``, which means the

Reviewer: ```suggestion
   *timeout* specifies the duration in seconds (as a :class:`float`)
   this method gives the executor to finish joining.
   With the default, ``None``, the
```

* Add key clarifying detail that timeout is in seconds as a `float`, which is never stated, and I had to dig into [the code](https://github.com/python/cpython/blob/5d09d11aa0b89aeba187f4f520728ccaf4fc5ac1/Lib/threading.py#L1094) to be sure. This leaves the reader to guess—Float seconds? Integer microseconds? `datetime` objects? `time` structs? Per Diataxis, there should be no doubt or ambiguity in reference.
* Be more concise, direct and to the point, as appropriate for reference docs

   method allows the executor an unlimited amount of time.

>    If the timeout duration is reached, this method emits a warning and

Reviewer: ```suggestion
   If the *timeout* is reached, this method emits a :exc:`RuntimeWarning` and
```

Explicitly specify and cross-reference _what_ warning is raised (extracted from the code), if users wanted to `except` it, filter it, check for it, log it, etc).

Also, format the parameter name accordingly

   terminates the default executor without waiting for its threads to finish
   joining.

   Note that there is no need to call this function when :func:`asyncio.run` is

>    used, because that high-level function handles default executor shutdown

Author: One spot where I think more explanation would be beneficial: explaining *why* "there is no need to call this function when `asyncio.run()` is used." Please correct my guess at the "why", if needed.

Author: @gvanrossum Is this "why" accurate?

Reviewer 3: Yeah, it's just because `asyncio.run()` calls various shutdown methods.

>    Note that there is no need to call this function when :func:`asyncio.run` is
>    used, because that high-level function handles default executor shutdown
>    automatically.

Reviewer: ```suggestion
   .. note:: There is no need to call this method when :func:`asyncio.run` is
      used, because that high-level function handles default executor shutdown
      automatically.
```

* Make this an actual note
* This is a method rather than a function, and calling it thus avoids repetition and potential confusion

Author: @CAM-Gerlach Is it common reST syntax to have the note text start on the same line as the `.. note::` directive? I'm pretty sure I've usually seen it always as:

```
.. note::

   There is no need ...
   ...
   ...
```

Reviewer: Stylistically, I typically do

```suggestion
   .. note::
   
      There is no need to call this method when :func:`asyncio.run` is
      used, because that high-level function handles default executor shutdown
      automatically.
```

But either way produces identical rendered output, so long as the directive is configured accordingly (as the default admonitions are).

.. coroutinemethod:: loop.shutdown_default_executor(timeout=None)

   Schedule the closure of the default executor and wait for it to join all of
   the threads in the :class:`~concurrent.futures.ThreadPoolExecutor`.
   After calling this method,
   :meth:`loop.run_in_executor` will raise a :exc:`RuntimeError` if called
   while using the default executor.

   *timeout* specifies the duration in seconds (as a :class:`float`)
   this method gives the executor to finish joining.
   With the default, ``None``, the
   method allows the executor an unlimited amount of time.

   If the *timeout* is reached, this method emits a :exc:`RuntimeWarning` and
   terminates the default executor without waiting for its threads to finish
   joining.

>    .. note::
>    
>       There is no need to call this method when :func:`asyncio.run` is
>       used, because that high-level function handles default executor shutdown
>       automatically.

Reviewer: ```suggestion
   .. note::

      There is no need to call this method when :func:`asyncio.run` is
      used, because that high-level function handles default executor shutdown
      automatically.
```

Seems like some trailing white space crept in to my previous example suggestion

.. coroutinemethod:: loop.shutdown_default_executor(timeout=None)

   Schedule the closure of the default executor and wait for it to join all of
   the threads in the :class:`~concurrent.futures.ThreadPoolExecutor`.
   After calling this method,
   :meth:`loop.run_in_executor` will raise a :exc:`RuntimeError` if called
   while using the default executor.

>    *timeout* specifies the duration in seconds (as a :class:`float`)

Reviewer 3: Call me old-fashioned, but I don't like starting a sentence with a lowercase letter, even if it's part of an identifier, and I don't like capitalizing the name either, so I liked the original phrasing:
```suggestion
   The *timeout* parameter specifies the duration in seconds (as a :class:`float`)
```

Reviewer: Yeah, I don't like it either, but on the other hand it was a little more concise and direct, as appropriate for an API reference per Diataxis. It just comes down which you want to prioritize...

Reviewer 3: @bskinn can choose then.

   Schedule the *callback* :term:`callback` to be called with
   *args* arguments at the next iteration of the event loop.

   Callbacks are called in the order in which they are registered. Each callback

>    will be called exactly once, even if it is registered multiple times.

Author: It wasn't clear to me why it was important to note that 'each callback will be called exactly once' -- if I register a callback once, I would expect it only to be run once. 

My guess here is that a callback could be registered multiple times before execution loops back around to it, and in this case it's still called only once?

Alternatively -- I could see a design decision where a callback, once registered, is called repeatedly as the event loop cycles around, up until the point where it is _de-registered_ by some mechanism... but that would be a more surprising design to me.

Regardless: I suggest more clarity here around this "will be called exactly once".

Author: @gvanrossum Can you provide guidance on why this 'will be called exactly once' language is here?

Reviewer 3: I don't recall, it's always been there even in the earliest version of the asyncio docs I can find.

Your addition "even if it is registered multiple times" is incorrect though -- if I register the exact same function twice, it will be called twice.

I recommend leaving this paragraph alone.

Reviewer 3: @bskinn
> if I register a callback once, I would expect it only to be run once.

That depends on the type of callback -- it is true here, but out of context that's a faulty assumption. E.g. `add_reader()` adds a callback that may be called multiple times.

Author: This change reverted entire by 7d18a15

   An optional keyword-only *context* argument allows specifying a
   custom :class:`contextvars.Context` for the *callback* to run in.
   Callbacks use the current context when no *context* is provided.

   Returns an instance of :class:`asyncio.Handle`, which can be
   used later to cancel the callback.

   This method is not thread-safe.

.. method:: loop.call_soon_threadsafe(callback, *args, context=None)

>    A thread-safe variant of :meth:`call_soon`. This method *must* be used to
>    schedule callbacks *from another thread*.

Reviewer: It was not totally clear to me whether this means to say that:
* This method must be used If _the callback_ will be run in another thread than _this method is called in_,, which is what I _think_ it is trying to say, or
* This method _must be used from another thread_ to schedule callbacks, as I first read it, or
* Something else?

Here's a suggestion trying to clarify the first option:

```suggestion
   A thread-safe variant of :meth:`call_soon`.
   This method **must** be used instead of :meth:`!call_soon` to
   schedule callbacks if they are to be called from another thread.
```

Also, used bold instead of italics to avoid confusion with the standard formatting of parameter names.

Author: @gvanrossum Can you provide clarity on the technical details here?

Reviewer 3: I'd simply reverse it: "When scheduling callbacks from another thread, this function must be called, since `call_soon()` is not thread-safe."

Reviewer: Taking the above and clarifying further the finer points a bit, specifically what "from another thread" refers to (another thread than the caller? The callback is executed in? Or something else?) and being more explicit that it must be used _instead of_ `call_soon`, both as above, we have

```suggestion
   A thread-safe variant of :meth:`call_soon`.
   When scheduling callbacks to be called from another thread,
   this method **must** be used instead of the non-thread-safe :meth:`!call_soon`.
```

Is this right?

Reviewer 3: Yeah, though I'm less excited than CAM about the **bold**. (I feel that **bold** for emphasis is pretty extreme, and I don't think there's confusion possible with other uses of italics here.

Author: Another option might be [RFC2199 "MUST"](https://www.rfc-editor.org/rfc/rfc2119)?

A quick search of the codebase shows what appear to be numerous uses of RFC2119, so this seems stylistically consistent to me.

Reviewer: To me that feels even stronger than the bold. If bold is too strong, then we can just use italics anyway; certainly not worth holding up this PR over:

```suggestion
   A thread-safe variant of :meth:`call_soon`.
   When scheduling callbacks to be called from another thread,
   this method *must* be used instead of the non-thread-safe :meth:`!call_soon`.
```

Author: Not to strain overly hard at a gnat, but --- to me, the emphasis of **bold** has a different meaning from the emphasis of MUST: Bold is a semantic emphasis, as part of an attempt to persuade the reader by reasoning; whereas MUST is a technical emphasis, indicating how one is to use the method correctly.

Given that it is critical in a technical sense for a caller to use this method and not `call_soon()` from another thread, I think the emphasis of MUST is specifically warranted here.

Reviewer: That's true, but I at least strongly (heh) associate it with mandatory clauses in normative specifications, as opposed to user-facing instructions. But...meh, it's not _that_ important to me what color we paint the bikeshed... :)

Author: I appreciate your perspective, it's what I was going for more than anything. (Apologies to anyone watching the PR for the noise as we discussed.)

On looking back at Guido's original recommendation, I think his wording is semantically different from your recommendation, so I'm actually going to edit to match his, since he is the subject matter expert here. Commit coming shortly.

>    An optional keyword-only *context* argument allows specifying a
>    custom :class:`contextvars.Context` for the *callback* to run in.

Reviewer 2: ```suggestion
   The optional keyword-only *context* argument specifies a
   custom :class:`contextvars.Context` for the *callback* to run in.
```

   Callbacks use the current context when no *context* is provided.

>    Returns an instance of :class:`asyncio.Handle`, which can be

Reviewer 2: ```suggestion
   Return an instance of :class:`asyncio.Handle`, which can be
```

Reviewer: Particularly if we frame it this way, this sentence/para would seem to fit best immediately after the first/summary line. This also follows the loose convention used other places in this module (e.g. in the following `call_later` method) and in other modules.
